include/albatross/src/models/patchwork_gp_details.hpp

```diff
@@ -0,0 +1,270 @@
+      std::is_same<decltype(nearest_group_test<GroupKey>(0)), GroupKey>::value;
+
+public:
+  static constexpr bool value =
+      has_valid_grouper && has_valid_boundary && has_valid_nearest_group;
+};
+
+} // namespace details
+
+/*
+ * A BoundaryFeature represents a pseudo observation of the difference
+ * between predictions from two different models.  In other words,
+ *
+ *   BoundaryFeature(key_i, key_j, feature)
+ *
+ * represents the quantity
+ *
+ *   model_i.predict(feature) - model_j.predict(feature)
+ *
+ * Patchwork Krigging uses these to force equivalence between two
+ * otherwise independent models.  These are the \delta_{k,l} variables
+ * in Equation 2 from the paper referenced above.
+ */
+template <typename GroupKey, typename FeatureType> struct BoundaryFeature {
+
+  BoundaryFeature(const GroupKey &left_key_, const GroupKey &right_key_,
+                  const FeatureType &feature_)
+      : left_key(left_key_), right_key(right_key_), feature(feature_){};
+
+  BoundaryFeature(GroupKey &&left_key_, GroupKey &&right_key_,
+                  FeatureType &&feature_)
+      : left_key(std::move(left_key_)), right_key(std::move(right_key_)),
+        feature(std::move(feature_)){};
+
+  GroupKey left_key;
+  GroupKey right_key;
+  FeatureType feature;
+};
+
+template <typename GroupKey, typename FeatureType>
+inline auto as_boundary_feature(GroupKey &&left_key, GroupKey &&right_key,
+                                FeatureType &&feature) {
+  using BoundaryFeatureType =
+      BoundaryFeature<typename std::decay<GroupKey>::type,
+                      typename std::decay<FeatureType>::type>;
+  return BoundaryFeatureType(std::forward<GroupKey>(left_key),
+                             std::forward<GroupKey>(right_key),
+                             std::forward<FeatureType>(feature));
+}
+
+template <typename GroupKey, typename FeatureType>
+inline auto as_boundary_features(GroupKey &&left_key, GroupKey &&right_key,
+                                 const std::vector<FeatureType> &features) {
+  using BoundaryFeatureType =
+      BoundaryFeature<typename std::decay<GroupKey>::type,
+                      typename std::decay<FeatureType>::type>;
+
+  std::vector<BoundaryFeatureType> boundary_features;
+  for (const auto &f : features) {
+    boundary_features.emplace_back(as_boundary_feature(left_key, right_key, f));
+  }
+  return boundary_features;
+}
+
+/*
+ * GroupFeature
+ *
+ * This is used to indicate which model a particular Feature corresponds
+ * to.  It corresponds (loosely) to the f_i in Equations 3 and 4 from
+ * the paper referenced above.
+ */
+
+template <typename GroupKey, typename FeatureType> struct GroupFeature {
+
+  GroupFeature() : key(), feature(){};
+
+  GroupFeature(const GroupKey &key_, const FeatureType &feature_)
+      : key(key_), feature(feature_){};
+
+  GroupFeature(GroupKey &&key_, FeatureType &&feature_)
+      : key(std::move(key_)), feature(std::move(feature_)){};
+
+  GroupKey key;
```

Comment:
Would it be more intuitive to rename `key` to simply `group`? I think that'd make the covariance functions more understandable.

Reply:
The problem with `group` in my mind is that it could be confused with the `value`.  In the vocabulary I've been using each "group" consists of a "key" and a "value".  Another option would be to change it to `group_key`, but that's getting pretty verbose (particularly in the other sections where it then becomes `left_group_key` and `right_group_key`.  thoughts on that last option?

Reply:
I think that is probably the way to go. I appreciate that it starts to get a little wordy, but I think it would help with understanding. After all, they are keys relating to groups and that can get forgotten easily if the variables don't indicate it.